### PR TITLE
fixes #7818, correct internal representation of generic objects array construction

### LINKED ISSUE #7818: inconsistent internal representation of generic objects array construction
the deeper I dig into #4799, the more I scratch my head. here is another meta-bug I think I need to isolate and fixed separately from #4799.

I know this one may looks like #7712 and #7637 because on surface it looks like object slicing. Indeed it sliced the object. But the problem highlighted here is the inconsistent internal representation of array construction of generic objects.

I will use macro to demonstrate the problem and to avoid object slicing at runtime. The problem itself not of macros module, but it originate from commonType algorithm of array construction semcheck.

```Nim
import macros

type
  Vehicle[T] = object of RootObj
    tire: T
  Car[T] = object of Vehicle[T]
  Bike[T] = object of Vehicle[T]

macro peek(n: typed): untyped =
  echo getTypeImpl(n).treeRepr
  echo "---------"

var v = Vehicle[int](tire: 3)
var c = Car[int](tire: 4)
var b = Bike[int](tire: 2)

peek([v, c, b])
peek([c, b, v]) # change elements position
```

output:
```text
BracketExpr
  Sym "array"
  Infix
    Ident ".."
    IntLit 0
    IntLit 2
  BracketExpr        # expected output: Vehicle[int]
    Sym "Vehicle"    
    Sym "int"
---------
BracketExpr
  Sym "array"
  Infix
    Ident ".."
    IntLit 0
    IntLit 2
  Sym "Vehicle"   # lost it's generic identity: Vehicle[?]
---------
````

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

File: compiler/sem.nim [hunk 1]
```diff
@@ -158,6 +158,8 @@ proc commonType*(x, y: PType): PType =
       b = b.lastSon.skipTypes({tyGenericInst})
     if a.kind == tyObject and b.kind == tyObject:
       result = commonSuperclass(a, b)
+      # bug #7818, put this into commonSuperclass?
+      if k in {tyRef, tyPtr}: result = skipTypes(result, skipPtrs)
```
Comment (Araq) [inline compiler/sem.nim]: This is still not correct. ``skipTypes`` skips every pointer indirection but the subtype relation only holds for a single indirection.

File: compiler/types.nim [hunk 2]
```diff
@@ -1056,8 +1056,13 @@ proc commonSuperclass*(a, b: PType): PType =
     x = x.sons[0]
   var y = b
   while y != nil:
+    var t = y # bug #7818, save type before skip
     y = skipTypes(y, skipPtrs)
-    if ancestors.contains(y.id): return y
+    if ancestors.contains(y.id):
+      # bug #7818, special case with tyRef and tyPtr
+      # skip it anyway, they don't play nice with tyGenericInst
+      if k in {tyRef, tyPtr}: t = t.skipTypes(skipPtrs)
```
Comment (jangko) [inline compiler/types.nim]: I move it here to make the code intention clear.
skipTypes in line 1064 is basically a deferred  skipTypes from line 1060. 
it is a special case with tyRef/tyPtr similar with skipTypes inside proc commonType line 157,158 of sem.nim

**Comment by jangko** [comment]
oh no, something really wrong here

File: compiler/types.nim [hunk 3]
```diff
@@ -1059,8 +1059,12 @@ proc commonSuperclass*(a, b: PType): PType =
     x = x.sons[0]
   var y = b
   while y != nil:
+    var t = y # bug #7818, save type before skip
     y = skipTypes(y, skipPtrs)
-    if ancestors.contains(y.id): return y
+    if ancestors.contains(y.id):
+      # bug #7818, defer the previous skipTypes
+      if k in {tyRef, tyPtr}: t = t.skipTypes(skipPtrs)
```
Comment (zah) [inline compiler/types.nim]: why not `t = y` here? No need to call `skipTypes` again arriving at the same result.

**Pull request merged**